### DISCUSSION TITLE
Make Yosys use the yosys-systemverilog (i.e. surelog+uhdm) plugin to process systemverilog files

### DISCUSSION THREAD

**Discussion by Andrea-Spitale** [The-OpenROAD-Project/OpenROAD-flow-scripts#1263] [Ideas]
Hello,
I've recently installed yosys + klayout + openroad + openroad flow scripts as alternative to openlane, as I wanted to modify yosys to process systemverilog files through surelog and uhdm, rather than sv2v.
To make this work in openroad flow, since the yosys commands are executed as tcl commands after importing them with the command "yosys - import" in scripts/synth_preamble.tcl, I've made the changes below to scripts/synth_preamble.tcl

(first install the yosys-systemverilog plugin from https://github.com/antmicro/yosys-systemverilog, following "Installation" section)
1) Add the command "plugin - i yosys_plugin/systemverilog.so" after "yosys -import" line (please mind that your installed plugin .so file may be located somewhere else)
2) modify the "read_verilog" command to become "yosys read_systemverilog" command, so that the newly added command is correctly executed through tcl.

Do you see any harm in the changes I've done? The flow seems to be working and I believe this info may help other users customize yosys within openroad flow. I don't know if this is something that could be added as a feature/option of the open road flow?

**Comment by maliberty** [comment]
It sounds reasonable.  I would make a good addition to ORFS.

**Reply by Andrea-Spitale** [reply]
Thanks for your kind reply. I forgot to mention that the script I modified is part of the openroad-flow-scripts folders, rather than open road application. I chose the wrong github project when opening the discussion

**Reply by vijayank88** [reply]
@vvbandeira FYI

**Reply by vijayank88** [reply]
@Andrea-Spitale It's fine. Both are maintained by same team.
Is possible to push/share code changes to refer the same?

**Reply by Andrea-Spitale** [reply]
Yes absolutely. Can I share the updated script somewhere or send it to the team, rather than pushing it to the repo?

**Reply by maliberty** [reply]
A pull request is easiest but you could also attach it here

**Comment by Andrea-Spitale** [comment]
Hello @maliberty  and @vijayank88 , here is the modified script. Please let me know if there is any issue.please mind that line 5 of the script has to be modified with the correct path, depending on where the user installs the plugin.
[synth_preamble_yosys_SV.zip](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/files/12113431/synth_preamble_yosys_SV.zip)

**Reply by gkamendje** [reply]
@Andrea-Spitale from what I can see, the script will not work if modules in your RTL code include Systemverilog packages located in other files. The following modification might be necessary in this case. Note the use of the `-defer` option and the additional call with the `-link` option.  
```
if {[info exist ::env(YOSYS_READ_SV)]} {
   foreach file $::env(VERILOG_FILES) {
      puts "reading file  $file"
      read_systemverilog -defer  -verbose -d 2 {*}$vIdirsArgs  -sv $file
   }
   read_systemverilog -link -verbose -d 2

} 
```

**Reply by gkamendje** [reply]
Having said that, there is still a long way to go to get `read_systemverilog` an integral part of the flow. Getting the `systemverilog_plugin` to work well on Centos7 is still a pain  see https://github.com/chipsalliance/systemverilog-plugin/issues/1934 and  https://github.com/chipsalliance/systemverilog-plugin/issues/1826.

**Reply by maliberty** [reply]
We are still supporting centos7 so this is a blocker.  I think next year we'll drop support as it reaches eol.

**Reply by gkamendje** [reply]
There have been some movement on the systemverilog plug-ing side. A recent code change in Yosys is preventing it from compiling with the latest version of Yosys but I think this should be fixed soon (https://github.com/chipsalliance/synlig/issues/1961). 

@maliberty is there a plan to support CentOS Stream 9 going forward? I was able to get ORFS to run on CentOS Stream 9. I would be more than happy to share the hacks.

**Reply by maliberty** [reply]
There is no plan to support centos in the future.  There is too much controversy about RH moves.  If the changes are no cost then we could take them but we won't ensure future compatibility.  We might replace centos with openSUSE but that's not decided.

**Comment by Andrea-Spitale** [comment]
@gkamendje that's interesting,thanks for your suggestion!
